File: 03-Azure/01-04-AI/01_Agentic_SDLC/src/api-cs/Controllers/OrderDetailDeliveriesController.cs
using Microsoft.AspNetCore.Mvc;
using OctoSupply.Api.Models;
using OctoSupply.Api.Repositories;
using OctoSupply.Api.Utils;

namespace OctoSupply.Api.Controllers;

[ApiController]
[Route("api/order-detail-deliveries")]
public sealed class OrderDetailDeliveriesController(OrderDetailDeliveriesRepository repository) : ControllerBase
{
    private readonly OrderDetailDeliveriesRepository _repository = repository;

    [HttpGet]
    public async Task<ActionResult<IReadOnlyList<OrderDetailDelivery>>> GetAll() =>
        Ok(await _repository.FindAllAsync());

    [HttpGet("{id:int}")]
    public async Task<ActionResult<OrderDetailDelivery>> GetById(int id)
    {
        var item = await _repository.FindByIdAsync(id);
        return item is null ? throw new NotFoundException("OrderDetailDelivery", id) : Ok(item);
    }

    [HttpPost]
    public async Task<ActionResult<OrderDetailDelivery>> Create([FromBody] CreateOrderDetailDeliveryRequest request)
    {
        if (request.OrderDetailId <= 0 || request.DeliveryId <= 0)
        {
            throw new ValidationException("orderDetailId and deliveryId are required");
        }

        var created = await _repository.CreateAsync(request);
        return CreatedAtAction(nameof(GetById), new { id = created.OrderDetailDeliveryId }, created);
    }

    [HttpPut("{id:int}")]
    public async Task<ActionResult<OrderDetailDelivery>> Update(int id, [FromBody] UpdateOrderDetailDeliveryRequest request) =>
        Ok(await _repository.UpdateAsync(id, request));

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Delete(int id)
    {
        await _repository.DeleteAsync(id);
        return NoContent();
    }
}

File: 03-Azure/01-04-AI/01_Agentic_SDLC/src/api-java/src/main/java/com/octocat/supply/controller/OrderDetailDeliveriesController.java
package com.octocat.supply.controller;

import com.octocat.supply.exception.NotFoundException;
import com.octocat.supply.model.OrderDetailDeliveryModel.CreateOrderDetailDeliveryRequest;
import com.octocat.supply.model.OrderDetailDeliveryModel.OrderDetailDelivery;
import com.octocat.supply.model.OrderDetailDeliveryModel.UpdateOrderDetailDeliveryRequest;
import com.octocat.supply.repository.OrderDetailDeliveriesRepository;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/order-detail-deliveries")
public class OrderDetailDeliveriesController {
    private final OrderDetailDeliveriesRepository repository;

    public OrderDetailDeliveriesController(OrderDetailDeliveriesRepository repository) {
        this.repository = repository;
    }

    @GetMapping
    public List<OrderDetailDelivery> getAll() {
        return repository.findAll();
    }

    @GetMapping("/{id}")
    public OrderDetailDelivery getById(@PathVariable int id) {
        return repository.findById(id).orElseThrow(() -> new NotFoundException("OrderDetailDelivery not found"));
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public OrderDetailDelivery create(@RequestBody CreateOrderDetailDeliveryRequest request) {
        ControllerValidation.requirePositive(request.orderDetailId(), "orderDetailId");
        ControllerValidation.requirePositive(request.deliveryId(), "deliveryId");
        return repository.create(request);
    }

    @PutMapping("/{id}")
    public OrderDetailDelivery update(@PathVariable int id, @RequestBody UpdateOrderDetailDeliveryRequest request) {
        repository.findById(id).orElseThrow(() -> new NotFoundException("OrderDetailDelivery not found"));
        repository.update(id, request);
        return repository.findById(id).orElseThrow(() -> new NotFoundException("OrderDetailDelivery not found"));
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void delete(@PathVariable int id) {
        if (!repository.delete(id)) {
            throw new NotFoundException("OrderDetailDelivery not found");
        }
    }
}
